Initial SDK functionality implemented

nbrl opened the pull request:
Currently:
* Allows automatic sending of touchpoint/sharepoint
* Manual sending of touchpoint/sharepoint with callbacks
* Registration of event listeners
* Calling of registered event listeners when an action occurs
* Storage of AT data in cookies/local storage


* Still spits out quite a lot of debugging data to the console. Leaving in for now.

jabclab commented on bower.json:
```diff
@@ -0,0 +1,15 @@
+  ],
+  "dependencies": {
+    "fingerprint": "~0.5.4",
+    "json2": "*"
```
Could we be more specific than `*`?

nbrl replied:
Yep, now referenced by commit hash.

jabclab commented on lib/cookieStorage.js:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Mozilla's cookies.js
+ * TODO: include at build time rather than including here.
```
This comment still applicable?

nbrl replied:
Remvoed.

jabclab commented on package.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "build-things-js-sdk",
+  "version": "0.0.0",
+  "description": "",
```
Would be worth adding a description here just for completeness :smile:

nbrl replied:
Done

jabclab commented:
A few comments lying around in `karma.conf.js`, are they still applicable?

jabclab commented on lib/sdk.js:
```diff
@@ -0,0 +1,539 @@
+        console.log(data);
+
+        var d = JSON.parse(JSON.stringify(data || {}));
+        if (!('_at' in d)) {
```
following may be slightly more explicit:

```js
if (!d.hasOwnProperty('_at')) {
```

nbrl replied:
`d.hasOwnProperty('_at')` will evaluate true if e.g. `d._at = undefined` and subsequent bits rely on it being an object.

jabclab commented on lib/sdk.js:
```diff
@@ -0,0 +1,539 @@
+        Sharepoint: {
+            name: 'Sharepoint',
+            // url: 'http://127.0.0.1:3000'
+            url: 'http://localhost:3001/campaign/token/sharepoint/data'
```
Likely don't want these pointing to `localhost`?

nbrl replied:
What do you suggest? The current Heroku apps? Ultimately it will need to be injected depending on $ENV used.

jabclab replied:
Let's go with the Heroku apps for now

jabclab commented on lib/sdk.js:
```diff
@@ -0,0 +1,539 @@
+        }
+
+        var duplicateData = false;
+        for (var i=0; i<currentStoreData[clientToken].length; i++) {
```
micro optimisation :wink: :

```
for (var i=0, len=currentStoreData[clientToken].length; i<len; i++) {
```

nbrl replied:
Done

jabclab commented on lib/sdk.js:
```diff
@@ -0,0 +1,539 @@
+     * back from the XHR is the same as the current one. When this occurs, a new
+     * token is requested.
+     * @param {string} name - Name of the triggered touchpoint.
+     * @param {object} d - Parsed JSON data object containing _at and _client.
```
`s/d/data/`

nbrl replied:
Why?

nbrl replied:
done

jabclab commented on lib/sdk.js:
```diff
@@ -0,0 +1,539 @@
+     * Manually send a touchpoint with the given name. The name is added to the
+     * _at object before sending using exactly the same call as used in init().
+     * @param {string} name - Name of the triggered touchpoint.
+     * @param {object} d - Parsed JSON data object containing _at and _client.
```
`s/d/data/`

nbrl replied:
done

jabclab commented on lib/sdk.js:
```diff
@@ -0,0 +1,539 @@
+     * send with the data object. If no such parameter is included, send tries
+     * to send the data as a touchpoint and a sharepoint.
+     */
+    function send(data, cb) {
```
No JSDoc for the params here

jabclab commented on bower.json:
```diff
@@ -0,0 +1,16 @@
+    "tests"
+  ],
+  "dependencies": {
+    "fingerprint": "~0.5.4",
```
be good to have a concrete version here

jabclab commented:
We'll put the `sdk.js` file in a (new) `src/` dir

jabclab commented on package.json:
```diff
@@ -0,0 +1,30 @@
+    "test": "./node_modules/karma/bin/karma start karma.conf.js"
+  },
+  "devDependencies": {
+    "expect.js": "^0.3.1",
```
Couple of deps missing in here, e.g. karma

jabclab commented:
Just need the licenses for the polyfills now :smile:

nbrl commented:
* `sdk.js` moved to `/src/sdk.js`
* removed temp. comments from `karma.conf`
* removed `*-old.js`
* polyfills are public domain, but added attribution and link anyway
* removed and reimplemented `cookieStorage.js` as this was using GPL'd code which does not allow inclusion Apache licensed projects. Now uses [js-coder/cookies.js](https://github.com/js-coder/cookie.js) (MIT)
